File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/polynomial/ejml/PolynomialFitEjml.java
package org.kevoree.modeling.api.polynomial.ejml;

import org.kevoree.modeling.api.polynomial.brain.Polynomial;

/**
 * Created by assaa_000 on 28/10/2014.
 */
public class PolynomialFitEjml implements Polynomial {

    // Vandermonde matrix
    DenseMatrix64F A;
    // matrix containing computed polynomial coefficients
    DenseMatrix64F coef;
    // observation matrix
    DenseMatrix64F y;

    // solver used to compute
    AdjLinearSolverQr solver;


    public PolynomialFitEjml(int degree) {
        coef = new DenseMatrix64F(degree+1,1);
        A = new DenseMatrix64F(1,degree+1);
        y = new DenseMatrix64F(1,1);

        // create a solver that allows elements to be added or removed efficiently
        solver = new AdjLinearSolverQr();
    }



    public double[] getCoef() {
        return coef.data;
    }


    public void fit( double samplePoints[] , double[] observations ) {
        y.reshape(observations.length,1,false);
        System.arraycopy(observations,0, y.data,0,observations.length);
        A.reshape(y.numRows, coef.numRows,false);

        // set up the A matrix
        for( int i = 0; i < observations.length; i++ ) {

            double obs = 1;

            for( int j = 0; j < coef.numRows; j++ ) {
                A.set(i,j,obs);
                obs *= samplePoints[i];
            }
        }

        // process the A matrix and see if it failed
        if( !solver.setA(A) )
            throw new RuntimeException("Solver failed");

        solver.solve(y,coef);
    }

}

File: org.kevoree.modeling.microframework.typescript/src/main/ts/org/kevoree/modeling/api/polynomial/util/PolynomialFitEjml.ts

class PolynomialFitEjml {

  public A: DenseMatrix64F = null;
  public coef: DenseMatrix64F = null;
  public y: DenseMatrix64F = null;
  public solver: AdjLinearSolverQr = null;

  constructor(degree: number) {
    this.coef = new DenseMatrix64F(degree + 1, 1);
    this.A = new DenseMatrix64F(1, degree + 1);
    this.y = new DenseMatrix64F(1, 1);
    this.solver = new AdjLinearSolverQr();
  }

  public getCoef(): number[] {
    return this.coef.data;
  }

  public fit(samplePoints: number[], observations: number[]): void {
    this.y.reshape(observations.length, 1, false);
    System.arraycopy(observations, 0, this.y.data, 0, observations.length);
    this.A.reshape(this.y.numRows, this.coef.numRows, false);
    for (var i: number = 0; i < observations.length; i++) {
      var obs: number = 1;
      for (var j: number = 0; j < this.coef.numRows; j++) {
        this.A.set(i, j, obs);
        obs *= samplePoints[i];
      }
    }
    this.solver.setA(this.A);
    this.solver.solve(this.y, this.coef);
  }

}

